src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -295,196 +339,211 @@ impl TcpPeer {
-            Some(Socket::Connecting { .. }) => return Poll::Ready(Err(Fail::new(EINPROGRESS, "socket connecting"))),
-            Some(Socket::Inactive { .. }) => return Poll::Ready(Err(Fail::new(EBADF, "socket inactive"))),
-            Some(Socket::Listening { .. }) => return Poll::Ready(Err(Fail::new(ENOTCONN, "socket listening"))),
-            None => return Poll::Ready(Err(Fail::new(EBADF, "bad queue descriptor"))),
-        };
-        match inner.established.get(&key) {
-            Some(ref s) => s.poll_recv(ctx),
-            None => Poll::Ready(Err(Fail::new(ENOTCONN, "connection not established"))),
+    pub fn poll_recv(&self, qd: QDesc, ctx: &mut Context) -> Poll<Result<DemiBuffer, Fail>> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::poll_recv");
+        let inner: Ref<Inner> = self.inner.borrow();
+        let mut qtable: RefMut<IoQueueTable<InetQueue>> = inner.qtable.borrow_mut();
+        match qtable.get_mut(&qd) {
+            Some(InetQueue::Tcp(ref mut queue)) => match queue.get_mut_socket() {
+                Socket::Established(ref mut socket) => socket.poll_recv(ctx),
+                Socket::Closing(ref mut socket) => socket.poll_recv(ctx),
+                Socket::Connecting(_) => Poll::Ready(Err(Fail::new(libc::EINPROGRESS, "socket connecting"))),
+                Socket::Inactive(_) => Poll::Ready(Err(Fail::new(libc::EBADF, "socket inactive"))),
+                Socket::Listening(_) => Poll::Ready(Err(Fail::new(libc::ENOTCONN, "socket listening"))),
+            },
+            _ => Poll::Ready(Err(Fail::new(libc::EBADF, "bad queue descriptor"))),
         }
     }
 
-    pub fn push(&self, fd: QDesc, buf: DemiBuffer) -> PushFuture {
-        let err: Option<Fail> = match self.send(fd, buf) {
+    pub fn push(&self, qd: QDesc, buf: DemiBuffer) -> PushFuture {
+        let err: Option<Fail> = match self.send(qd, buf) {
             Ok(()) => None,
             Err(e) => Some(e),
         };
-        PushFuture { fd, err }
+        PushFuture { qd, err }
     }
 
-    pub fn pop(&self, fd: QDesc) -> PopFuture {
+    pub fn pop(&self, qd: QDesc) -> PopFuture {
         PopFuture {
-            fd,
+            qd,
             inner: self.inner.clone(),
         }
     }
 
-    fn send(&self, fd: QDesc, buf: DemiBuffer) -> Result<(), Fail> {
-        let inner = self.inner.borrow_mut();
-        let key = match inner.sockets.get(&fd) {
-            Some(Socket::Established { local, remote }) => (*local, *remote),
-            Some(..) => return Err(Fail::new(ENOTCONN, "connection not established")),
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        };
-        match inner.established.get(&key) {
-            Some(ref s) => s.send(buf),
-            None => Err(Fail::new(ENOTCONN, "connection not established")),
+    fn send(&self, qd: QDesc, buf: DemiBuffer) -> Result<(), Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::send");
```

Comment:
Could we add changes that concern performance observation in a different commit?